ts/lib/commands/precompile.ts

```diff
@@ -0,0 +1,72 @@
+import os from 'os';
+import execa from 'execa';
+import fs from 'fs-extra';
+import path from 'path';
+import { command } from '../utilities/ember-cli-entities';
+import copyDeclarations from '../utilities/copy-declarations';
+
+export const PRECOMPILE_MANIFEST = 'dist/.ts-precompile-manifest';
+
+export default command({
+  name: 'ts:precompile',
+  works: 'insideProject',
+  description:
+    'Generates JS and declaration files from TypeScript sources in preparation for publishing.',
+
+  availableOptions: [{ name: 'manifest-path', type: String, default: PRECOMPILE_MANIFEST }],
+
+  async run(options: { manifestPath: string }) {
+    let outDir = `${os.tmpdir()}/e-c-ts-precompile-${process.pid}`;
+    let { paths, rootDir, pathRoots } = this._loadConfig(outDir);
+    if (!paths) {
+      this.ui.writeLine('No `paths` were found in your `tsconfig.json`, so `ts:precompile` is a no-op.');
+      return;
+    }
+
+    // prettier-ignore
+    await execa('tsc', [
+      '--declarationDir', outDir,
+      '--rootDir', rootDir || this.project.root,
+      '--allowJs', 'false',
+      '--noEmit', 'false',
+      '--declaration',
+      '--emitDeclarationOnly',
+      '--sourceMap', 'false',
+      '--inlineSourceMap', 'false',
+      '--inlineSources', 'false',
+    ]);
+
+    let manifestPath = options.manifestPath;
+    let packageName = this.project.pkg.name;
+    let createdFiles = copyDeclarations(pathRoots, paths, packageName, this.project.root);
+
+    fs.mkdirsSync(path.dirname(manifestPath));
+    fs.writeFileSync(manifestPath, JSON.stringify(createdFiles.reverse()));
+    fs.remove(outDir);
+  },
+
+  _loadConfig(outDir: string) {
+    let ts = this.project.require('typescript') as typeof import('typescript');
+    let configPath = ts.findConfigFile(this.project.root, ts.sys.fileExists);
+    if (!configPath) {
+      throw new Error('Unable to locate `tsconfig.json`');
+    }
+
+    let configSource = ts.readJsonConfigFile(configPath, ts.sys.readFile);
+    let config = ts.parseJsonSourceFileConfigFileContent(configSource, ts.sys, path.dirname(configPath));
```

Comment:
It's a little difficult to identify the correct way to read config files. 

https://github.com/mike-north/code-to-json/blob/a1395ce8c478dfc00856e10cbba11f711e057640/packages/cli/src/commands/run.ts#L76-L78 worked for me. Some subset of the approaches I tried failed to handle `{ extends: "foo" }` properly, so that's probably worth testing

Reply:
This properly follows `extends` chains, and I generally prefer it over `readConfigFile` as it has better type guarantees (you get a `ParsedCommandLine` rather than `any`).